### PR TITLE
Create base VWSException for erroneous responses from VWS

### PR DESCRIPTION
**Pull request by laymonage**

### CODE REVIEW

**Comment by adamtheturtle** [comment]
Thank you @laymonage for this draft PR. Please ping me when you would like a review!

**Comment by adamtheturtle** [comment]
I know this is a draft, but would a user not expect `UnknownVWSErrorPossiblyBadName` to be a `VWSException`?

**Comment by laymonage** [comment]
@adamtheturtle I wasn't sure if we should only use `VWSException` for exceptions that are in the [table](https://library.vuforia.com/content/vuforia-library/en/articles/Solution/How-To-Use-the-Vuforia-Web-Services-API.html#How-To-Interperete-VWS-API-Result-Codes), but you're right, I guess it's better if we just make all of them inherit from `VWSException` to simplify things.

**Comment by laymonage** [comment]
I based `VWSException` from `BaseVWSException` to distinguish exceptions that have a `Response` object as property and those that don't. If that's not what you want, we can remove `BaseVWSException` and just use `VWSException`, and set an empty `Response()` (or `None`) as the default value.

File: src/vws/exceptions.py [hunk 1]
```diff
@@ -18,80 +40,33 @@ class UnknownVWSErrorPossiblyBadName(Exception):
     """
 
 
-class ConnectionErrorPossiblyImageTooLarge(
-    requests.exceptions.ConnectionError,
```
Comment (adamtheturtle) [inline src/vws/exceptions.py:22]: Was there value derived from inheriting from `requests.exceptions.ConnectionError` which is now lost?

**Comment by adamtheturtle** [comment]
Thank you again @laymonage for your continued work on this.

> @adamtheturtle I wasn't sure if we should only use `VWSException` for exceptions that are in the [table](https://library.vuforia.com/content/vuforia-library/en/articles/Solution/How-To-Use-the-Vuforia-Web-Services-API.html#How-To-Interperete-VWS-API-Result-Codes), but you're right, I guess it's better if we just make all of them inherit from `VWSException` to simplify things.

In this draft PR I see that `MatchProcessing` and `MaxNumResultsOutOfRange` are made to inherit from `VWSException`.

You will notice that in the table you linked to, there is no `MatchProcessing` or `MaxNumResultsOutOfRange`.

There is a naming confusion in this package which stems from Vuforia changing and mixing names. Vuforia Web Services (VWS) mostly refers to the APIs which allow one to add and manage targets. The Vuforia Web Query API (VWQ) allows one to make 

I appreciate that this problem is present in the naming of this tool being `vws-python` and with the exceptions file and the package being `vws`.

Perhaps we could have `VWQException` or `CloudRecoException`.
This may mean that someone has to do `except (VWSException, CloudRecoException)` but I think I'm OK with that.
This also resolves the "problem" that `ConnectionErrorPossiblyImageTooLarge` is inherently different to the VWS Exceptions from the table.

> I based `VWSException` from `BaseVWSException` to distinguish exceptions that have a `Response` object as property and those that don't. If that's not what you want, we can remove `BaseVWSException` and just use `VWSException`, and set an empty `Response()` (or `None`) as the default value.

I think that as a user it is hard to know whether I should use `VWSException` or `BaseVWSException`.

Looking at your comment on https://github.com/VWS-Python/vws-python/issues/1271#issuecomment-674513881 I see:

> I want to catch the exceptions that might arise from this package, but I don't want to handle them individually.

This is not possible with this change as it is with `BaseVWSException` and `VWSException` if I want to rely on `response` when it is there. I have to use:

```python
try:
    ... # vws-specific code
except BaseVWSException:
    ...
except VWSException as exc:
    status_code = exc.response.status_code
    ...
```

I think that this is confusing and I have to carefully inspect the documentation to understand what is `BaseVWSException` and what is plain `VWSException`. I'm happy to discuss if you disagree.


In the current code base, the following is the only place in which `UnknownVWSErrorPossiblyBadName` is raised:

https://github.com/VWS-Python/vws-python/blob/f3356f9e5629229d8b2834a655affb53123c01c6/src/vws/_result_codes.py#L48-L52

We can see here that `response` exists - it just is not passed through to `UnknownVWSErrorPossiblyBadName.__init__`.

I believe that you can therefore make the change to have `UnknownVWSErrorPossiblyBadName` take a `response` and have a `response` property.

This does not resolve your motivation completely as it still leaves `ConnectionErrorPossiblyImageTooLarge`, but it resolves it for all exceptions raised by Vuforia Web Services (not the Vuforia Web Query API).

------

In summary, I believe that there are a few options and a few issues.
My guess of the best way forward is:

* Add `response` to `UnknownVWSErrorPossiblyBadName`.
* Instead of `BaseVWSException` and `VWSException`, go with `CloudRecoException` and `VWSException`, or just exclude the `CloudReco` exceptions from this unification.
* Add a test which would fail if the changes in this PR were reverted (this could be modifications of existing tests).

File: docs/source/exceptions.rst [hunk 2]
```diff
@@ -3,3 +3,4 @@ Exceptions
 
 .. automodule:: vws.exceptions
    :members:
+   :show-inheritance:
```
Comment (adamtheturtle) [inline docs/source/exceptions.rst:6]: Thank you.
Please also use `:inherited-members:` so that a reader can know all the properties of each exception without following the inheritance tree.

File: src/vws/exceptions.py [hunk 3]
```diff
@@ -5,11 +5,33 @@
...
 from requests import Response
 
 
-class UnknownVWSErrorPossiblyBadName(Exception):
+class BaseVWSException(Exception):
+    """Base class for Vuforia exceptions."""
+
+
+class VWSException(BaseVWSException):
+    """Exception that has the Vuforia response as a property."""
```
Comment (adamtheturtle) [inline src/vws/exceptions.py:16]: In this codebase we use:

```python
"""
...
"""
```

rather than:

```python
"""..."""
```

even for single line docstrings.

File: src/vws/exceptions.py (hunk 1 above)
Comment (laymonage) [reply src/vws/exceptions.py:22]: No, you [raised the `ConnectionErrorPossiblyImageTooLarge` class](https://github.com/VWS-Python/vws-python/blob/master/src/vws/query.py#L125), so Python instantiates it without any arguments. If you want to retain the request and response object in the exception, you would have to do:

```py
raise ConnectionErrorPossiblyImageTooLarge(request=exc.request, response=exc.response) from exc
```

**Comment by adamtheturtle** [comment]
This is looking really good @laymonage - please let me know when you would like me to have a look.

**Comment by laymonage** [comment]
Hi @adamtheturtle, I think it's ready for review.

**Comment by adamtheturtle** [comment]
Thank you @laymonage ! In particular for your persistence on this.

I appreciate your work and your high quality code. I will merge this and follow up with some minor changes - for example by crediting you in the `CHANGELOG` and doing a release.

I will also consider moving some things around - perhaps `vws.vws_exceptions/vwq_exceptions/custom_exceptions` or similar.

File: tests/test_exceptions.py [hunk 4]
```diff
@@ -333,3 +345,71 @@ def test_match_processing(
...
+    A VWSException or CloudRecoException should have a response property
+    and string representation with the text property of the response.
+    """
+    base_classes: List[Union[Type[CloudRecoException], Type[VWSException]]] = [
+        CloudRecoException,
+        VWSException,
+    ]
+    for base in base_classes:
+        response = requests.Response()
+        setattr(response, '._content', bytes(f'Test{base.__name__}', 'ascii'))
```
Comment (laymonage) [inline tests/test_exceptions.py:412]: Thanks, FYI I missed something here, it should be `'_content'` instead of `'._content'`.
